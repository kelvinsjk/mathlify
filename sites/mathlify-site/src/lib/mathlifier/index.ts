/**
 * generates djot string,
 * using JavaScript's template literal syntax to handle mathematical expressions (both static and dynamic)
 *
 * math: starts with ${}, terminates with \n,
 * display: starts with $${}, terminates with \n\n,
 * display env: starts with $${'align'}, etc.
 * text: starts with @${}, terminates immediately
 */
export function mathlify(strings: TemplateStringsArray, ...values: unknown[]): string {
	return mathlifyGen(modules)(strings, ...values);
}

const modules: Modules = {
	math: (x: string) => `$\`${x}\``,
	display,
	mathEnvs: {
		equation: (x: string) => display(insertEnv('equation', x)),
		align: (x: string) => display(insertEnv('align', x)),
		gather: (x: string) => display(insertEnv('gather', x)),
		alignat: (x: string) => display(insertEnv('alignat', x)),
		equationStar: (x: string) => display(insertEnv('equation*', x)),
		alignStar: (x: string) => display(insertEnv('align*', x)),
		gatherStar: (x: string) => display(insertEnv('gather*', x)),
		alignatStar: (x: string) => display(insertEnv('alignat*', x)),
	},
};

function insertEnv(env: string, content: string, args = ''): string {
	if (env === 'alignat' || env === 'alignat*') {
		let firstLine = content.split('\n')[0];
		if (!firstLine) {
			firstLine = content.split('\n')[1];
		}
		const twoN = firstLine.split('&').length;
		if (twoN % 2 !== 0)
			console.warn(`alignat should have an even number of columns, but got ${twoN} instead`);
		args = `{${Math.floor(twoN / 2)}}`;
	}
	// todo: handle this with regex
	while (content.endsWith('\n') || content.endsWith('\t') || content.endsWith(' ')) {
		content = content.slice(0, content.length - 1);
	}
	return `\\begin{${env}}${args}${content}\n\\end{${env}}`;
}

function display(x: string): string {
	return `$$\`${x}\``;
}

enum Modes {
	text, // 0
	math, // 1
	display, // 2
	mathEnv, // 3
}

export interface Modules {
	math: (x: string) => string;
	display: (x: string) => string;
	mathEnvs?: {
		equation?: (x: string) => string;
		align?: (x: string) => string;
		gather?: (x: string) => string;
		alignat?: (x: string) => string;
		equationStar?: (x: string) => string;
		alignStar?: (x: string) => string;
		gatherStar?: (x: string) => string;
		alignatStar?: (x: string) => string;
	};
}

enum MathEnvs {
	equation,
	equationStar,
	align,
	alignStar,
	gather,
	gatherStar,
	alignat,
	alignatStar,
}

const mathEnvsKeys = [
	'equation',
	'equation*',
	'align',
	'align*',
	'gather',
	'gather*',
	'alignat',
	'alignat*',
] as const;
function isMathEnv(x: string): x is (typeof mathEnvsKeys)[number] {
	return (mathEnvsKeys as readonly string[]).includes(x);
}

const envToMode: Record<(typeof mathEnvsKeys)[number], MathEnvs> = {
	equation: MathEnvs.equation,
	'equation*': MathEnvs.equationStar,
	align: MathEnvs.align,
	'align*': MathEnvs.alignStar,
	gather: MathEnvs.gather,
	'gather*': MathEnvs.gatherStar,
	alignat: MathEnvs.alignat, // triggered by $${'alignat'}{x}
	'alignat*': MathEnvs.alignatStar, // triggered by $${'alignat'}{x}
};

interface MathEnvOptions {
	mathEnv: MathEnvs;
	eqnColumns?: number;
}

// TODO: alignat cols

export function mathlifyGen(
	modules: Modules,
): (strings: TemplateStringsArray, ...values: unknown[]) => string {
	return (strings: TemplateStringsArray, ...values: unknown[]) => {
		let currentModeString = '';
		let output = '';
		let mode = Modes.text;
		let mathEnvOptions: MathEnvOptions = { mathEnv: MathEnvs.equation };
		strings.forEach((str, i) => {
			const nextVal = values[i] ?? '';
			if (mode === Modes.text) {
				if (str.endsWith('@')) {
					// continues text mode
					currentModeString += `${str.slice(0, str.length - 1)}${nextVal}`;
				} else {
					// starts new text environment
					[mode, output, currentModeString, mathEnvOptions] = startNewEnv(
						str,
						'',
						nextVal,
						currentModeString,
					);
				}
			} else if (mode === Modes.math) {
				// checks for \n or \r\n non-greedily
				const regex = /([^]*?)(\r?\n)([^]*)/;
				const match = str.match(regex);
				if (match) {
					// end math mode
					const [, before, newline, after] = match;
					output += `${before}`;
					if (output) {
						currentModeString += modules.math(output);
					}
					[mode, output, currentModeString, mathEnvOptions] = startNewEnv(
						after,
						newline,
						nextVal,
						currentModeString,
					);
				} else {
					// continue math mode
					if (str.endsWith(`@`)) {
						// TODO: check this
						output += `${str.slice(0, str.length - 1)}`;
					} else {
						output += `${str}${nextVal}`;
					}
					if (i === strings.length - 1) {
						// final string
						currentModeString += modules.math(output);
					}
				}
			} else if (mode === Modes.display) {
				// checks for \n\n or \r\n\r\n within str. group into (#1)(#2) where #2 = \n or \r\n and it is done non-greedily
				const regex = /([^]*?)(\r?\n[ \t]*\r?\n)([^]*)/;
				const match = str.match(regex);
				if (match) {
					// end display mode
					const [, before, newline, after] = match;
					output += before;
					if (output) {
						currentModeString += modules.display(output);
					}
					[mode, output, currentModeString, mathEnvOptions] = startNewEnv(
						after,
						newline,
						nextVal,
						currentModeString,
					);
				} else {
					// continue display mode
					if (str.endsWith(`@`)) {
						// TODO: check this
						output += `${str.slice(0, str.length - 1)}`;
					} else {
						output += `${str}${nextVal}`;
					}
					if (i === strings.length - 1) {
						// final string
						currentModeString += modules.display(output);
					}
				}
			} else if (mode === Modes.mathEnv) {
				// checks for \n\n or \r\n\r\n within str. group into (#1)(#2) where #2 = \n or \r\n and it is done non-greedily
				const regex = /([^]*?)(\r?\n[ \t]*\r?\n)([^]*)/;
				const match = str.match(regex);
				if (match) {
					// end mathEnv mode
					const [, before, newline, after] = match;
					output += before;
					if (output) {
						currentModeString += insertMathEnv(mathEnvOptions, output, modules);
					}
					[mode, output, currentModeString, mathEnvOptions] = startNewEnv(
						after,
						newline,
						nextVal,
						currentModeString,
					);
				} else {
					// continue mathEnv mode
					if (str.endsWith(`@`)) {
						// TODO: check this.
						output += `${str.slice(0, str.length - 1)}`;
					} else {
						output += `${str}${nextVal}`;
					}
					if (i === strings.length - 1) {
						// final string
						currentModeString += insertMathEnv(mathEnvOptions, output, modules);
					}
				}
			}
		});
		return currentModeString;
	};
}

// returns [Mode, acc, curr, {mathEnv, eqnColumns?}]
function startNewEnv(
	after: string,
	newline: string,
	nextVal: unknown,
	curr: string,
): [Modes, string, string, MathEnvOptions] {
	const defaultMathEnvOptions = { mathEnv: MathEnvs.equation };
	let mathEnvOptions: MathEnvOptions = defaultMathEnvOptions;
	if (after.endsWith('$')) {
		let x = `${nextVal}`;
		let mode = Modes.display;
		if (isMathEnv(x)) {
			// new math env
			mode = Modes.mathEnv;
			mathEnvOptions = { mathEnv: envToMode[x] };
			//TODO: handle alignat cols
			x = '';
		}
		// new display mode
		return [mode, x, curr + `${newline}${after.slice(0, after.length - 1)}`, mathEnvOptions];
	} else if (after.endsWith('@')) {
		// new text mode
		return [
			Modes.text,
			'',
			curr + `${newline}${after.slice(0, after.length - 1)}${nextVal}`,
			defaultMathEnvOptions,
		];
	} else {
		// new math mode
		return [Modes.math, `${nextVal}`, curr + newline + after, defaultMathEnvOptions];
	}
}

function insertMathEnv(mathEnvOptions: MathEnvOptions, acc: string, modules: Modules): string {
	const mathEnv = mathEnvOptions.mathEnv;
	if (mathEnv === MathEnvs.equation) {
		return modules.mathEnvs?.equation ? modules.mathEnvs.equation(acc) : acc;
	} else if (mathEnv === MathEnvs.equationStar) {
		return modules.mathEnvs?.equationStar ? modules.mathEnvs.equationStar(acc) : acc;
	} else if (mathEnv === MathEnvs.align) {
		return modules.mathEnvs?.align ? modules.mathEnvs.align(acc) : acc;
	} else if (mathEnv === MathEnvs.alignStar) {
		return modules.mathEnvs?.alignStar ? modules.mathEnvs.alignStar(acc) : acc;
	} else if (mathEnv === MathEnvs.gather) {
		return modules.mathEnvs?.gather ? modules.mathEnvs.gather(acc) : acc;
	} else if (mathEnv === MathEnvs.gatherStar) {
		return modules.mathEnvs?.gatherStar ? modules.mathEnvs.gatherStar(acc) : acc;
	} else if (mathEnv === MathEnvs.alignat) {
		return modules.mathEnvs?.alignat ? modules.mathEnvs.alignat(acc) : acc;
	} else if (mathEnv === MathEnvs.alignatStar) {
		return modules.mathEnvs?.alignatStar ? modules.mathEnvs.alignatStar(acc) : acc;
	}
	console.warn(`mathEnv ${mathEnv} not found`);
	return acc;
}
