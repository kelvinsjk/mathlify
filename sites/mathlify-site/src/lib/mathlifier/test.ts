import { mathlify } from './';

const str = mathlify`This is inline math: ${'x+2'}xyz${'k'}
and this is text.

This is display math: $${'x+2'}xyz
abcd
${'k'}

This is aligned:
$${'align'}
  x+2 &= ${'k'} \\\\
  y &= ${'z'}`;

console.log(str);
