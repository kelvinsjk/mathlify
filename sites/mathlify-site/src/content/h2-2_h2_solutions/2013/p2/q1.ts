import { Answer } from '$lib/classes/answer';
import { fnTerm, Polynomial, quotient } from 'mathlify';
import { generateDomain } from '$content/h2-1_h2_learn/01_functions/02_inverse/02_domain/02_practice';
import { compositeExists } from '$content/h2-1_h2_learn/01_functions/03_composite/01_existence/02_practice';
import { compositeFormula } from '$content/h2-1_h2_learn/01_functions/03_composite/02_formula/02_practice';
import { EquationWorking } from 'mathlify/working';
import { mathlifierDj as mathlifier } from 'mathlifier';
import { Interval } from '$content/h2-1_h2_learn/01_functions/_intervals';
import { QED } from '$lib/utils/typesetting/utils';

export const answer = new Answer();

// ! Complete
// f(x) = (2+x)/(1-x)
// g(x) = 1-2x

const [a, b, c] = [-1, 1, 2];
const ascending = true;
const f = quotient(new Polynomial([c, b], { ascending }), new Polynomial([1, -1], { ascending }));
const g = new Polynomial([1, -2], { ascending });
const rhs = 5;
const fState = {
	fnType: 'improper',
	a,
	b,
	c,
	restriction: false,
	unknownConstants: false
} as const;
const gState = {
	fnType: 'linear',
	a: -2,
	b: 1,
	c: 0,
	restriction: false,
	unknownConstants: false
} as const;
const compositeStateFg: Parameters<typeof compositeExists>[0] = {
	f: fState,
	g: gState,
	fg: true
};

// a
{
	const { ans, soln } = compositeExists(compositeStateFg, [f, g]);
	answer.addPart(ans, soln);
}
// b
{
	const {
		ans: ans1,
		soln: soln1,
		exp: gf
	} = compositeFormula([f, generateDomain(fState), g, [Interval.ALL_REAL]], false, false, {
		noDomain: true,
		ansInline: true,
		combineFraction: true,
		QED: true
	});

	const gfx = fnTerm('gf', 'x');
	const working = new EquationWorking(gfx, rhs, { aligned: true });
	working.subIn({ gf }).crossMultiply({ hide: true }).expand();
	const { root } = working.solve.linear('x');
	const soln =
		soln1 +
		mathlifier`$${'align*'} \\text{Let } \\quad x &= (gf)^{-1}(${rhs})  \\\\ ${working}
	
	$${{}} (gf)^{-1} (${rhs}) = ${root} \\; ${QED}`;
	const ans = ans1 + '\\\n' + mathlifier`${{}} (gf)^{-1} (${rhs}) = ${root}.`;
	answer.addPart(ans, soln);
}
