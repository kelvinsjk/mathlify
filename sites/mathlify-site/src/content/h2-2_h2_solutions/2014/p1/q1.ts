import { Answer } from '$lib/classes/answer';
import { mathlifierDj as mathlifier } from 'mathlifier';
import { Expression, Polynomial } from 'mathlify';
import { fractionalInverseWorking } from '$content/h2-1_h2_learn/01_functions/02_inverse/03_formula/02_practice';
import { compositeFormula } from '$content/h2-1_h2_learn/01_functions/03_composite/02_formula/02_practice';
import { QED } from '$lib/utils/typesetting/utils';

export const answer = new Answer();

// ! Complete
// 1/(1-x)

const f = new Expression([1, '/', new Polynomial([1, -1], { ascending: true })]);

// a
{
	const { soln: f2Working, exp: f2 } = compositeFormula([f, [], f, []], true, false, {
		gName: 'f',
		noDomain: true
	});
	const { working: fInvWorking } = fractionalInverseWorking(f, false, { reportInverse: true });
	const soln = f2Working + fInvWorking + mathlifier`Hence ${'f^2(x) = f^{-1}(x)'} \\; ${QED}`;
	const ans = mathlifier`${{}}f^2(x)=f^{-1}(x) = ${f2}.`;
	answer.addPart(ans, soln);
}
// b
{
	const soln = mathlifier`$${'align*'} f^3(x) &= f^2f(x) \\\\ &= f^{-1}f(x) \\quad \\text{since } f^2(x)=f^{-1}(x) \\\\ &= x \\; ${QED}`;
	const ans = mathlifier`${{}}f^3(x)=x.`;
	answer.addPart(ans, soln);
}
