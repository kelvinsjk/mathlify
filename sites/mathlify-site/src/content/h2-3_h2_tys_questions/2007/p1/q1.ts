/* The question here is the copyright of 
- Singapore Examinations and Assessment Board
- Cambridge Assessment International Education
- Ministry of Education, Singapore
- Government of Singapore

We provide this question on Mathlify under fair use for archival purposes.
These are questions more than 10 years old and are difficult to obtain
commercially.

For the newer questions, please purchase a copy of the
"ten year series" from authorized publishers and distributors.
*/

import { Polynomial, quotient, sum } from 'mathlify';
import { combineFraction } from 'mathlify/algebra';
import { Question } from '$lib/classes/question';
import { mathlifierDj as mathlifier } from 'mathlifier';

export const question = new Question();

// (2x^2 - x - 19) / (x^2 + 3x + 2)
// (x^2 - 4x - 21) / (x^2 + 3x + 2)

const num = new Polynomial([2, -1, -19]);
const den = new Polynomial([1, 3, 2]);
const frac = quotient(num, den);
const secondTerm = 1;
const lhs = sum(frac, -secondTerm);
const rhs = combineFraction(lhs);

// a
{
	question.addPart(
		mathlifier`Show that

$${lhs} = ${rhs}.`,
		{ marks: 1 }
	);
}
// b
{
	question.addPart(
		mathlifier`Hence, without using a calculator, solve the inequality

$${frac} > ${secondTerm}.`,
		{ marks: 4 }
	);
}
