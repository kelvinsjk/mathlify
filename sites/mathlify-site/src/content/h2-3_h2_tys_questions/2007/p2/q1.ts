/* The question here is the copyright of 
- Singapore Examinations and Assessment Board
- Cambridge Assessment International Education
- Ministry of Education, Singapore
- Government of Singapore

We provide this question on Mathlify under fair use for archival purposes.
These are questions more than 10 years old and are difficult to obtain
commercially.

For the newer questions, please purchase a copy of the
"ten year series" from authorized publishers and distributors.
*/

import { Question } from '$lib/classes/question';
import { mathlifierDj as mathlifier } from 'mathlifier';

export const question = new Question();

const matrix = [
	[1.15, 0.6, 0.55, 8.28],
	[1.2, 0.45, 0.3, 6.84],
	[2.15, 0.9, 0.65, 13.05]
];
const leeLian = [1.3, 0.25, 0.5];
const rows = ['| :-- | :---: | :---: | :---: | :---: |', '| | Suresh | Fandi | Cindy | Lee Lian |'];
const headings = ['Pineapples (kg)', 'Mangoes (kg)', 'Lychees (kg)', 'Total amount paid in \\$'];
for (let i = 0; i < 4; i++) {
	rows.push(
		`| ${headings[i]} | ${matrix[0][i].toFixed(2)} | ${matrix[1][i].toFixed(2)} | ${matrix[2][i].toFixed(2)} | ${leeLian[i]?.toFixed(2) ?? ''} |`
	);
}
const table = rows.join('\n');

question.addBody(
	mathlifier`Four friends buy three different kinds of fruit in the market. When they get home they cannot remember
the individual prices per kilogram, but three of them can remember the total amount that they each
paid.

The weights of fruit and the total amounts paid are shown in the following table.

@${table}

Assuming that, for each variety of fruit, the price per kilogram paid by each of the friends is the same,
calculate the total amount that Lee Lian paid.`,
	{ marks: 6 }
);
