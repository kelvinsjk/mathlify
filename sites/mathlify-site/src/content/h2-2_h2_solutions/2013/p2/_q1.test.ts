import { expect, it } from 'vitest';
import { answer } from './q1';
import { mathlifierDj as mathlifier } from 'mathlifier';

const ans = answer.answer;
const actual: typeof ans = {
	parts: [
		{
			body: mathlifier`${{}} fg
does not exist as ${{}} {R_{g} \\not \\subseteq D_{f}.}`
		},
		{
			body: mathlifier`${{}} gf(x) = \\frac{-3-3x}{1-x}.

${{}} (gf)^{-1}(5) = 4.`
		}
	]
};

it('2013p2q1', () => {
	expect(answer).toMatchSnapshot();
	expect(ans).toEqual(actual);
});
