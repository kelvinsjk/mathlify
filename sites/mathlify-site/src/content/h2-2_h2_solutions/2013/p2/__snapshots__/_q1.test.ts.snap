// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`2013p2q1 1`] = `
Answer {
  "answer": {
    "parts": [
      {
        "body": "$\` fg \`
does not exist as $\`  {R_{g} \\not  \\subseteq D_{f}.} \`",
      },
      {
        "body": "$\` gf(x) = \\frac{- 3 - 3x}{1 - x}. \`\\
$\`  (gf)^{-1} (5) = 4. \`",
      },
    ],
  },
  "solution": {
    "parts": [
      {
        "body": "$$\` \\begin{align*}
R_{g} &= \\left( -\\infty, \\infty \\right) \\\\
D_{f} &= \\left( -\\infty, 1 \\right) \\cup \\left( 1, \\infty \\right)
\\end{align*} \`

Since $\`  R_{g} \\not  \\subseteq D_{f}, \`
the composite function $\` fg \`
does not exist $\` \\QED  \`
",
      },
      {
        "body": "$$\` \\begin{align*}
gf(x) &= g \\left( \\frac{2 + x}{1 - x} \\right)
\\\\ &= 1 - 2\\left( \\frac{2 + x}{1 - x} \\right)
	\\\\ &= 1 - \\frac{2\\left( 2 + x \\right)}{1 - x}
	\\\\ &= 1 - \\frac{4 + 2x}{1 - x}
	\\\\ &= \\frac{- 3 - 3x}{1 - x} \\; \\QED
\\end{align*} \`$$\` \\begin{align*} \\text{Let } \\quad x &= (gf)^{-1}(5)  \\\\ gf\\left(x\\right) &= 5
	\\\\ \\frac{- 3 - 3x}{1 - x} &= 5
	\\\\ - 3 - 3x &= 5 - 5x
	\\\\ 2x &= 8
	\\\\ x &= 4
\\end{align*} \`
	
	$$\`  (gf)^{-1} (5) = 4 \\; \\QED  \`",
      },
    ],
  },
}
`;
